arch/xtensa/src/esp32s3/esp32s3_irq.c

```diff
@@ -0,0 +1,686 @@
+
+/****************************************************************************
+ * Public Data
+ ****************************************************************************/
+
+/* g_current_regs[] holds a reference to the current interrupt level
+ * register storage structure.  It is non-NULL only during interrupt
+ * processing.  Access to g_current_regs[] must be through the macro
+ * CURRENT_REGS for portability.
+ */
+
+volatile uint32_t *g_current_regs[1];
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+/* Maps a CPU interrupt to the IRQ of the attached peripheral interrupt */
+
+static uint8_t g_cpu0_intmap[ESP32S3_NCPUINTS];
+
+static volatile uint8_t g_irqmap[NR_IRQS];
+
+/* g_intenable[] is a shadow copy of the per-CPU INTENABLE register
+ * content.
+ */
+
+static uint32_t g_intenable[1];
+
+/* Bitsets for free, unallocated CPU interrupts available to peripheral
+ * devices.
+ */
+
+static uint32_t g_cpu0_freeints = ESP32S3_CPUINT_PERIPHSET;
+
+/* Bitsets for each interrupt priority 1-5 */
+
+static const uint32_t g_priority[5] =
+{
+  ESP32S3_INTPRI1_MASK,
+  ESP32S3_INTPRI2_MASK,
+  ESP32S3_INTPRI3_MASK,
+  ESP32S3_INTPRI4_MASK,
+  ESP32S3_INTPRI5_MASK
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: esp32s3_intinfo
+ *
+ * Description:
+ *    Return the CPU interrupt map of the given CPU and the register map
+ *    of the given peripheral.
+ *
+ ****************************************************************************/
+
+static void esp32s3_intinfo(int cpu, int periphid,
+                            uintptr_t *regaddr, uint8_t **intmap)
+{
+  *regaddr = CORE0_MAP_REGADDR(periphid);
+  *intmap  = g_cpu0_intmap;
+}
+
+/****************************************************************************
+ * Name:  esp32s3_getcpuint
+ *
+ * Description:
+ *   Get a free CPU interrupt for a peripheral device.  This function will
+ *   not ignore all of the pre-allocated CPU interrupts for internal
+ *   devices.
+ *
+ * Input Parameters:
+ *   intmask - mask of candidate CPU interrupts.  The CPU interrupt will be
+ *             be allocated from free interrupts within this set
+ *
+ * Returned Value:
+ *   On success, a CPU interrupt number is returned.
+ *   A negated errno is returned on failure.
+ *
+ ****************************************************************************/
+
+static int esp32s3_getcpuint(uint32_t intmask)
+{
+  uint32_t *freeints;
+  uint32_t bitmask;
+  uint32_t intset;
+  int cpuint;
+  int ret = -ENOMEM;
+  int cpu = 0;
+
+  /* Check if there are CPU interrupts with the requested properties
+   * available.
+   */
+
+  cpu = up_cpu_index();
```

Comment:
ESP32-S3 is a dual core chip, so yes we do expect an index other than `0`.  However, for this initial support all SMP features are now ignored.  These will be dealt with later.
Regarding `volatile uint32_t *g_current_regs[1];` this will change to the usual:
```
#ifdef CONFIG_SMP
/* For the case of architectures with multiple CPUs, then there must be one
 * such value for each processor that can receive an interrupt.
 */

volatile uint32_t *g_current_regs[CONFIG_SMP_NCPUS];

#else

volatile uint32_t *g_current_regs[1];

#endif
```

Reply:
Recently we had https://github.com/apache/incubator-nuttx/pull/5264 so we can go with `volatile uint32_t *g_current_regs[CONFIG_SMP_NCPUS];` even right now. But this is minor comment. It is fine to go with `volatile uint32_t *g_current_regs[1];` for now and rework later